src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.GeneratedFileReplacingCompilationTracker.cs

```diff
@@ -0,0 +1,190 @@
+                    newCompilation = underlyingCompilation.ReplaceSyntaxTree(existingSyntaxTree, newSyntaxTree);
+                }
+                else
+                {
+
+                    // The existing output no longer exists in the underlying compilation. This could happen if the user made
+                    // an edit which would cause this file to no longer exist, but they're still operating on an open representation
+                    // of that file. To ensure that this snapshot is still usable, we'll just add this document back in. This is not a
+                    // semantically correct operation, but working on stale snapshots never has that guarantee.
+                    newCompilation = underlyingCompilation.AddSyntaxTrees(newSyntaxTree);
+                }
+
+#pragma warning restore CS8774 // Member must have a non-null value when exiting.
+
+                Interlocked.CompareExchange(ref _compilationWithReplacement, newCompilation, null);
+            }
+
+            public bool HasCompilation => _underlyingTracker.HasCompilation;
+
+            public ProjectState ProjectState => _underlyingTracker.ProjectState;
+
+            public ICompilationTracker Clone()
+            {
+                throw new NotImplementedException();
+            }
+
+            public bool ContainsAssemblyOrModuleOrDynamic(ISymbol symbol, bool primary)
+            {
+                throw new NotImplementedException();
+            }
+
+            public bool? ContainsSymbolsWithNameFromDeclarationOnlyCompilation(Func<string, bool> predicate, SymbolFilter filter, CancellationToken cancellationToken)
+            {
+                throw new NotImplementedException();
+            }
+
+            public bool? ContainsSymbolsWithNameFromDeclarationOnlyCompilation(string name, SymbolFilter filter, CancellationToken cancellationToken)
+            {
+                throw new NotImplementedException();
+            }
+
+            public ICompilationTracker Fork(ProjectState newProject, CompilationAndGeneratorDriverTranslationAction? translate = null, bool clone = false, CancellationToken cancellationToken = default)
+            {
+                throw new NotImplementedException();
+            }
+
+            public ICompilationTracker FreezePartialStateWithTree(SolutionState solution, DocumentState docState, SyntaxTree tree, CancellationToken cancellationToken)
+            {
+                // Because we override SourceGeneratedDocument.WithFrozenPartialSemantics directly, we shouldn't be able to get here.
+                throw ExceptionUtilities.Unreachable;
+            }
+
+            public async Task<Compilation> GetCompilationAsync(SolutionState solution, CancellationToken cancellationToken)
+            {
+                await EnsureNewCompilationComputedAsync(solution, cancellationToken).ConfigureAwait(false);
+                return _compilationWithReplacement;
```

Comment:
Contract that this is nonnull?

Reply:
Rewritten to avoid this pattern. (In the original form EnsureNewCompilationComputedAsync was annotated so this was being checked by the compiler.)